### PR TITLE
feat(oui-datagrid): allow to keep selection across pages

### PR DESCRIPTION
**Pull request by marie-j**
### Description of the Change

Add parameter to allow to keep row selection between datagrid pages

### CODE REVIEW

File: packages/oui-datagrid/README.md [hunk 1]
```diff
@@ -127,6 +127,38 @@ Clicked row action 1: <span ng-if="$ctrl.action1Row">{{$ctrl.action1Row.lastName
   When <code class="oui-doc-codespan">selectable-rows</code> is provided, you can listen to checkboxes changes with the <code class="oui-doc-codespan">on-row-select</code> event. It provides <strong>$row</strong> as the triggered row and <strong>$rows</strong> as the complete list of rows selected so far.
 </oui-message>
 
+#### Keep selection across pages 
+
+```html:preview
+<oui-datagrid rows="$ctrl.data" page-size="5" selectable-rows on-row-select="$ctrl.onRowSelect($row, $rows)" keep-selected-rows>
+  <oui-column title="'First name'" property="firstName" sortable="asc"></oui-column>
```
Comment (FredericEspiau) [inline packages/oui-datagrid/README.md]: Maybe you could simplify the example, so we can see directly that what matters is in the `oui-datragrid` element :)

File: packages/oui-datagrid/README.md [hunk 2]
```diff
@@ -127,6 +127,38 @@ Clicked row action 1: <span ng-if="$ctrl.action1Row">{{$ctrl.action1Row.lastName
   When <code class="oui-doc-codespan">selectable-rows</code> is provided, you can listen to checkboxes changes with the <code class="oui-doc-codespan">on-row-select</code> event. It provides <strong>$row</strong> as the triggered row and <strong>$rows</strong> as the complete list of rows selected so far.
 </oui-message>
 
+#### Keep selection across pages 
+
+```html:preview
+<oui-datagrid rows="$ctrl.data" page-size="5" selectable-rows on-row-select="$ctrl.onRowSelect($row, $rows)" keep-selected-rows>
```
Comment (FredericEspiau) [inline packages/oui-datagrid/README.md]: Shouldn't `keep-selected-rows` be the default option for version 3 ?

Also, `selectable-rows`, `on-row-select` and `keep-selected-rows` don't look alike at all but they all concern the same feature, maybe you could standardize the naming ? `selectable-rows-retaining` ?

File: packages/oui-datagrid/src/cell/cell.controller.js [hunk 3]
```diff
@@ -24,7 +24,7 @@ export default class {
             this._compileCell();
         }
 
-        this.cellScope.$watch(() => this.datagridCtrl.selectedRows[this.index], (isSelected) => {
+        this.cellScope.$watch(() => this.datagridCtrl.currentSelectedRows[this.index], (isSelected) => {
```
Comment (FredericEspiau) [inline packages/oui-datagrid/src/cell/cell.controller.js]: `current` meaning in the current page ?

File: packages/oui-datagrid/src/datagrid.controller.js [hunk 4]
```diff
@@ -321,18 +324,20 @@ export default class DatagridController {
...
-            if (isSelected) {
-                result.push(this.displayedRows[index]);
-            }
-            return result;
-        }, []);
+        return this.selectedRows;
     }
 
     toggleRowSelection (index, isSelected) {
+        const row = this.displayedRows[index];
```
Comment (FredericEspiau) [inline packages/oui-datagrid/src/datagrid.controller.js]: Use either `displayed` or `current` for all variables related to the things on the current page :)

Comment (FredericEspiau) [reply packages/oui-datagrid/src/datagrid.controller.js]: (I like `displayed` better, `current` doesn't really make sense)

File: packages/oui-datagrid/src/datagrid.controller.js [hunk 5]
```diff
@@ -361,6 +368,17 @@ export default class DatagridController {
         });
     }
 
+    applyRowsSelection () {
+        if (this.keepSelectedRows) {
+            this.currentSelectedRows = this.displayedRows.map(row => !!find(this.selectedRows, selectedRow => isEqual(selectedRow, row)));
+            this.selectAllRows = this.currentSelectedRows.filter(isSelected => isSelected).length === this.displayedRows.length;
```
Comment (FredericEspiau) [inline packages/oui-datagrid/src/datagrid.controller.js]: `allRowsAreSelected` or `allRowsSelected` ?

File: packages/oui-datagrid/src/datagrid.controller.js [hunk 6]
```diff
@@ -361,6 +368,17 @@ export default class DatagridController {
...
+        } else {
+            this.selectedRows = [];
+            this.currentSelectedRows = this.currentSelectedRows.map(() => false);
```
Comment (FredericEspiau) [inline packages/oui-datagrid/src/datagrid.controller.js]: Both `currentSelectedRows` and `selectAllRows` are always updated at the same time. This is code smell

File: packages/oui-datagrid/src/index.spec.js [hunk 7]
```diff
@@ -518,6 +518,42 @@ describe("ouiDatagrid", () => {
                 element.scope().$apply();
                 expect(element.find("oui-datagrid-extra-top").text()).toBe("5");
             });
+
+            it("should keep selection across pages", () => {
+                const element = TestUtils.compileTemplate(`
+                <oui-datagrid rows="$ctrl.rows" page-size="5" keep-selected-rows>
```
Comment (FredericEspiau) [inline packages/oui-datagrid/src/index.spec.js]: No `selectable-rows` ?

File: packages/oui-datagrid/src/index.spec.js [hunk 8]
```diff
@@ -518,6 +518,42 @@ describe("ouiDatagrid", () => {
...
+                ctrl.toggleRowSelection(0, true);
+                getNextPagePaginationButton(element).triggerHandler("click");
+                element.scope().$apply();
+
+                const selected = ctrl.getSelectedRows();
+                expect(selected.length).toBe(1);
+                expect(selected[0]).toEqual(fakeData[0]);
+            });
+
+            it("should keep global selection across pages", () => {
```
Comment (FredericEspiau) [inline packages/oui-datagrid/src/index.spec.js]: There is no concept of `global selection` in the code, maybe the `selectedRows` should be named `globallySelectedRows` :)

File: packages/oui-datagrid/src/index.spec.js [hunk 9]
```diff
@@ -518,6 +518,42 @@ describe("ouiDatagrid", () => {
...
+            it("should keep selection across pages", () => {
+                const element = TestUtils.compileTemplate(`
+                <oui-datagrid rows="$ctrl.rows" page-size="5" keep-selected-rows>
+                    <oui-column property="firstName"></oui-column>
+                    <oui-column property="lastName"></oui-column>
+                </oui-datagrid>`, {
+                    rows: fakeData.slice(0, 10)
+                });
+                const ctrl = element.controller("ouiDatagrid");
+                ctrl.toggleRowSelection(0, true);
```
Comment (FredericEspiau) [inline packages/oui-datagrid/src/index.spec.js]: you should be clicking on the first checkbox instead

File: packages/oui-datagrid/src/index.spec.js [hunk 10]
```diff
@@ -518,6 +518,42 @@ describe("ouiDatagrid", () => {
...
+                </oui-datagrid>`, {
+                    rows: fakeData.slice(0, 10)
+                });
+                const ctrl = element.controller("ouiDatagrid");
+                ctrl.toggleAllRowsSelection(true);
+                getNextPagePaginationButton(element).triggerHandler("click");
+                element.scope().$apply();
+
+                const selected = ctrl.getSelectedRows();
+                expect(selected.length).toBe(5);
```
Comment (FredericEspiau) [inline packages/oui-datagrid/src/index.spec.js]:
```suggestion
                 expect(selected.length).toBe(pageSize);
```

File: packages/oui-datagrid/src/index.spec.js [hunk 11]
```diff
@@ -518,6 +518,42 @@ describe("ouiDatagrid", () => {
...
+                element.scope().$apply();
+
+                const selected = ctrl.getSelectedRows();
+                expect(selected.length).toBe(1);
+                expect(selected[0]).toEqual(fakeData[0]);
+            });
+
+            it("should keep global selection across pages", () => {
+                const element = TestUtils.compileTemplate(`
+                <oui-datagrid rows="$ctrl.rows" page-size="5" keep-selected-rows>
```
Comment (FredericEspiau) [inline packages/oui-datagrid/src/index.spec.js]:
```suggestion
                 <oui-datagrid rows="$ctrl.rows" page-size="${pageSize}" keep-selected-rows>
```

File: packages/oui-datagrid/src/index.spec.js (hunk 10 above)
Comment (FredericEspiau) [reply packages/oui-datagrid/src/index.spec.js]: This way what you want is explicit :)

File: packages/oui-datagrid/src/index.spec.js (hunk 8 above)
Comment (marie-j) [reply packages/oui-datagrid/src/index.spec.js]: Does `All rows` selection sounds better to you ?

**Pull request closed without merging**